sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/UploadUtils.java

```diff
@@ -39,72 +37,25 @@
         ParallelTransferOptions parallelTransferOptions,
         final Function<Flux<ByteBuffer>, Mono<Response<T>>> uploadInChunks,
         final BiFunction<Flux<ByteBuffer>, Long, Mono<Response<T>>> uploadFull) {
-        final long[] bufferedDataSize = {0};
-        final LinkedList<ByteBuffer> cachedBuffers = new LinkedList<>();
 
-        /*
-         * Window the reactive stream until either the stream completes or the windowing size is hit. If the window
-         * size is hit the next emission will be the pointer to the rest of the reactive steam.
-         *
-         * Once the windowing has completed buffer the two streams, this should create a maximum overhead of ~4MB plus
-         * the next stream emission if the window size was hit. If there are two streams buffered use Stage Blocks and
-         * Put Block List as the upload mechanism otherwise use Put Blob.
-         */
+        PayloadSizeGate gate = new PayloadSizeGate(parallelTransferOptions.getMaxSingleUploadSizeLong());
+
         return data
             .filter(ByteBuffer::hasRemaining)
-            .windowUntil(buffer -> {
-                if (bufferedDataSize[0] > parallelTransferOptions.getMaxSingleUploadSize()) {
-                    return false;
+            .concatMap(gate::write)
```

Comment:
Can you add some comments here describing the behavior:
The gate will only start emitting data after the threshold has been breached once it has been flushed. Checking the state of the threshold on receiving the first buffer means we can infer what upload operation we should do.

Reply:
added comments